File: src/quest_editor/gui/EditorHistory.ts
import { Disposable } from "../../core/observable/Disposable";
import { editor, IPosition, KeyCode, KeyMod } from "monaco-editor";
import { Disposer } from "../../core/observable/Disposer";
import IStandaloneCodeEditor = editor.IStandaloneCodeEditor;
import ICursorPositionChangedEvent = editor.ICursorPositionChangedEvent;
import IEditorMouseEvent = editor.IEditorMouseEvent;
import ScrollType = editor.ScrollType;

export class EditorHistory implements Disposable {
    private readonly history: IPosition[] = [];
    private history_index = -1;
    private capture_history = true;
    private readonly disposer = new Disposer();

    constructor(private readonly editor: IStandaloneCodeEditor) {
        this.disposer.add_all(
            this.editor.onDidChangeCursorPosition(this.did_change_cursor_position),

            this.editor.addAction({
                id: "phantasmal.action.back",
                label: "Back",
                keybindings: [KeyMod.Alt | KeyCode.LeftArrow],
                run: this.back,
            }),

            this.editor.addAction({
                id: "phantasmal.action.forward",
                label: "Forward",
                keybindings: [KeyMod.Alt | KeyCode.RightArrow],
                run: this.forward,
            }),

            this.editor.onMouseUp(this.mouse_up),
        );
    }

    dispose(): void {
        this.disposer.dispose();
    }

    reset(): void {
        this.history.splice(0, Infinity);
        this.history_index = -1;
    }

    private did_change_cursor_position = (e: ICursorPositionChangedEvent): void => {
        if (!this.capture_history) return;

        this.history.splice(this.history_index + 1, Infinity);

        if (
            e.source === "api" ||
            this.history_index === -1 ||
            Math.abs(e.position.lineNumber - this.history[this.history_index].lineNumber) >= 10
        ) {
            this.history.push(e.position);
            this.history_index++;
        } else {
            this.history[this.history_index] = e.position;
        }
    };

    private back = async (): Promise<void> => {
        if (this.history_index > 0) {
            this.set_position(this.history[--this.history_index]);
        }
    };

    private forward = async (): Promise<void> => {
        if (this.history_index + 1 < this.history.length) {
            this.set_position(this.history[++this.history_index]);
        }
    };

    private set_position = (position: IPosition): void => {
        this.capture_history = false;
        this.editor.setPosition(position);
        this.editor.revealPositionInCenterIfOutsideViewport(position, ScrollType.Immediate);
        this.capture_history = true;
    };

    private mouse_up = (e: IEditorMouseEvent): void => {
        const button = e.event.browserEvent.button;
        const buttons = e.event.browserEvent.buttons;

        if (button === 3) {
            if (buttons === 0) {
                e.event.preventDefault();
                this.back();
            }

            this.editor.focus();
        } else if (button === 4) {
            if (buttons === 0) {
                e.event.preventDefault();
                this.forward();
            }

            this.editor.focus();
        }
    };
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/asm/monaco/EditorHistory.kt
package world.phantasmal.web.questEditor.asm.monaco

import world.phantasmal.web.externals.monacoEditor.*
import world.phantasmal.webui.DisposableContainer
import kotlin.math.abs

class EditorHistory(private val editor: IStandaloneCodeEditor) : DisposableContainer() {
    private val history = mutableListOf<IPosition>()
    private var historyIndex = -1
    private var captureHistory = true

    init {
        addDisposables(
            editor.onDidChangeCursorPosition(::onDidChangeCursorPosition).toDisposable(),

            editor.addAction(object : IActionDescriptor {
                override var id = "phantasmal.action.back"
                override var label = "Go Back"
                override var keybindings = arrayOf(KeyMod.Alt or KeyCode.LeftArrow)

                override fun run(editor: ICodeEditor, vararg args: dynamic) {
                    goBack()
                }
            }).toDisposable(),

            editor.addAction(object : IActionDescriptor {
                override var id = "phantasmal.action.forward"
                override var label = "Go Forward"
                override var keybindings = arrayOf(KeyMod.Alt or KeyCode.RightArrow)

                override fun run(editor: ICodeEditor, vararg args: dynamic) {
                    goForward()
                }
            }).toDisposable(),

            editor.onMouseUp(::onMouseUp).toDisposable(),

            editor.onDidChangeModel(::reset).toDisposable(),
        )
    }

    private fun reset() {
        history.clear()
        historyIndex = -1
    }

    private fun onDidChangeCursorPosition(e: ICursorPositionChangedEvent) {
        if (!captureHistory) return

        while (history.lastIndex > historyIndex) {
            history.removeLast()
        }

        if (
            e.source === "api" ||
            historyIndex == -1 ||
            abs(e.position.lineNumber - history[historyIndex].lineNumber) >= 10
        ) {
            history.add(e.position.unsafeCast<IPosition>())
            historyIndex++
        } else {
            history[historyIndex] = e.position.unsafeCast<IPosition>()
        }
    }

    private fun goBack() {
        if (historyIndex > 0) {
            setPosition(history[--historyIndex])
        }
    }

    private fun goForward() {
        if (historyIndex < history.lastIndex) {
            setPosition(history[++historyIndex])
        }
    }

    private fun setPosition(position: IPosition) {
        captureHistory = false
        editor.setPosition(position)
        editor.revealPositionInCenterIfOutsideViewport(position, ScrollType.Immediate)
        captureHistory = true
    }

    private fun onMouseUp(e: IEditorMouseEvent) {
        val button = e.event.browserEvent.button.toInt()
        val buttons = e.event.browserEvent.buttons.toInt()

        if (button == 3) {
            if (buttons == 0) {
                e.event.preventDefault()
                goBack()
            }

            editor.focus()
        } else if (button == 4) {
            if (buttons == 0) {
                e.event.preventDefault()
                goForward()
            }

            editor.focus()
        }
    }
}
